### DISCUSSION TITLE
Exploring a multitenant pocketbase build...again :)

### DISCUSSION THREAD

**Discussion by benallfree** [pocketbase/pocketbase#3504] [Q&A] [answered]
I'm looking at doing a major overhaul of PocketHost in prep for v1.0, and I'm considering switching to Go because we learned a lot from the previous versions of PocketHost and are in need of some architectural changes to support multiple regions, etc. It would be a good time to switch, and probably the last chance. The "multitenant backend in 1 file" approach is showing a lot of appeal, too.

If I used PocketBase as a library and made a Go proxy, I'm curious to know if I could instantiate multiple instances of the PocketBase server on different ports, but in the same Go process. Can you think of any technical issues with that? I recall something about goja not liking coroutines, but as long as it's different instances of goja would that be okay?

The alternative is to write a Go proxy that launches separate Docker containers with PocketBase. That's what the nodejs version does now, and it achieves great isolation, security, and resource management. But perhaps at the expense of performance.

Anyway, just curious to know if running multiple PocketBase objects is even an option.

**Comment by ganigeorgiev** [comment, marked as answer]
The goja integration shouldn't cause issues in this case because we serialize each handler function (middleware, route, hook, etc.) into a string and execute it in its own `goja.Runtime` (_some general overview of this approach could be found in https://github.com/pocketbase/pocketbase/discussions/2825#discussioncomment-6376210_).

Technically you can start multiple PocketBase applications in a single Go process, each in their own goroutine, but you'll have to manually initialize each PocketBase instance and manually call [`apis.Serve(app, config)`](https://pkg.go.dev/github.com/pocketbase/pocketbase@v0.18.10/apis#Serve). The only problem with this that I can think of at the moment is the auto migration runner because we currently use a "shared" `migrate.MigrationsList` variables that cannot be changed from the Serve config (_I'll consider maybe allow customizing it_) - https://github.com/pocketbase/pocketbase/blob/1527b5ea4fc56e264b1f146687b660520b5bc9a3/apis/serve.go#L220-L229 meaning that all PocketBase instances will share the same migrations.

Another thing that also you have to keep in mind is that `pb_hooks` and `pb_migrations` scripts can be loaded only during the application startup, meaning that you'll have to attach a listener for the directories that will take care to restart (destroy and reinitialize) only the specific PocketBase instance.

But even if the above wasn't an issue, for your use case I don't think this will be suitable approach because you still will need some kind of isolation between the instances, either full blown VMs or containers, because users from different PocketBase applications will be able to access the other apps resources since they can execute shell commands within the `pb_hooks`.

**Comment by benallfree** [comment]
Related questions, since I'm in Go anyway...

I'd like to customize the PocketBase server to shut down after N seconds of no open requests (including SSE) and no active background tasks such as backups are happening. Do you have thoughts on how?

**Reply by ganigeorgiev** [reply]
I thought a little more on this but I'm not sure what to recommend as this may not be so easy to implement because we don't track the background tasks (aka. running goroutines) or even the active connections (_the latter is not that difficult_).

I guess you can check with [`runtime.NumGoroutine`](https://pkg.go.dev/runtime#NumGoroutine) with a `time.Timer` but I'm not sure how reliable that will be because you'll always have couple goroutines active (and their number can change in the future). For example, currently a new goroutine is fired on:
- each new request
- when sending emails from a public endpoint (password, verification, etc. )with the "fire and forget" calls
- when auto deleting files from deleted records (_we perform "optimistic" delete to avoid blocking the delete transaction in case the deleted records have a lot of files or S3 is used_)
- various concurrent "grouping" operations (eg. with [`errgroup`](https://pkg.go.dev/golang.org/x/sync/errgroup))
- and probably others...


To check whether there is a backup/restore operation you can call `app.Cache().Has(core.CacheKeyActiveBackup)`.

To track connections I guess you can either create a small struct with 2 fields - int and `sync.Mutex` OR maybe use the [`sync/atomic`](https://pkg.go.dev/sync/atomic) package. The active connections state for example can be tracked by attaching to the `OnBeforeServe` hook and using [e.Server.ConnState](https://pkg.go.dev/net/http#Server.ConnState) callback. 

Here is a short example (_I haven't tested it so treat the below as pseudo-code_):

```go
package main

import (
	"log"
	"net"
	"net/http"
	"sync/atomic"

	"github.com/pocketbase/pocketbase"
	"github.com/pocketbase/pocketbase/core"
	"github.com/pocketbase/pocketbase/tools/cron"
)

var connectionsCount atomic.Int64

func main() {
	app := pocketbase.New()

	app.OnBeforeServe().Add(func(e *core.ServeEvent) error {
		// check on every 5 minutes
		// for more fine grained ticker you can use time.Timer and check its channel
		idleScheduler := cron.New()
		idleScheduler.Add("idleCheck", "*/5 * * * *", func() {
			hasActiveConnections := connectionsCount.Load() > 0
			hasActiveBackup := app.Cache().Has(core.CacheKeyActiveBackup)

			if !hasActiveBackup && !hasActiveConnections {
				if err := e.Server.Close(); err == http.ErrServerClosed {
					idleScheduler.Stop()
					return // no more checks
				} else {
					log.Println(err) // failed to shutdown
				}
			}
		})
		idleScheduler.Start()

		e.Server.ConnState = func(c net.Conn, cs http.ConnState) {
			// for all connection state types check https://pkg.go.dev/net/http#ConnState
			switch cs {
			case http.StateNew:
				connectionsCount.Add(1)
			case http.StateClosed, http.StateHijacked:
				// this should restart the ticker after each connection
				connectionsCount.Add(-1)
				idleScheduler.Start()
			}
		}

		return nil
	})

	if err := app.Start(); err != nil {
		log.Fatal(err)
	}
}
```

**Reply by benallfree** [reply]
That's probably about as perfect as we can get, thanks for thinking it through!

**Comment by benallfree** [comment]
How would I pipe stdout/stderr to `os.Stdout` and `os.Stderr`?

```go
			app := pocketbase.New()
			app.Bootstrap()
			exited := make(chan bool, 1)
			go func() {
				apis.Serve(app, apis.ServeConfig{
					HttpAddr: fmt.Sprintf("127.0.0.1:%d", port),
				})
				exited <- true
			}()

			fmt.Printf("Instance launched on port %d\n", port)
			<-exited
```

**Reply by benallfree** [reply]
Or maybe it doesn't output the normal banner?

**Reply by ganigeorgiev** [reply]
I'm not sure I follow. By default the Go logger should already prints to stderr.

> Or maybe it doesn't output the normal banner?

Yes, you can set [`ServeConfig.ShowStartBanner`](https://pkg.go.dev/github.com/pocketbase/pocketbase@v0.18.10/apis#ServeConfig) to true if you want the default startup message.

**Comment by benallfree** [comment]
Following up...

`ConnState` seemed to be picking up other socket connections, not just request lifecycles. I ended up with this, do you see any issues with this approach?

```go
func addIdleWatchdog(app *pocketbase.PocketBase, timeout int64) {
	var connectionsCount atomic.Int64

	app.OnBeforeServe().Add(func(e *core.ServeEvent) error {

		arm, cancel := util.Deadman(func(rearm util.RearmFunc) {
			hasActiveConnections := connectionsCount.Load() > 0
			hasActiveBackup := app.Cache().Has(core.CacheKeyActiveBackup)

			util.Log("Has active backup: %t / hasActiveConnections: %t", hasActiveBackup, hasActiveConnections)

			if hasActiveBackup || hasActiveConnections {
				rearm()
				return
			}

			err := e.Server.Close()
			util.Log(err.Error()) // failed to shutdown
		})
		arm(timeout)

		requestLogger := func(next echo.HandlerFunc) echo.HandlerFunc {
			return func(c echo.Context) error {
				start := time.Now()
				util.Log("Request started")
				connectionsCount.Add(1)
				cancel()

				// Call the next handler (which is your main handler)
				err := next(c)

				// This code will run after your handler has finished
				util.Log("Request ended (took %v)\n", time.Since(start))
				connectionsCount.Add(-1)
				arm(timeout)
				return err
			}
		}

		e.Router.Use(requestLogger)

		return nil
	})
}
```

**Reply by ganigeorgiev** [reply]
Yes, I guess using a middleware should also suffice for your use case, but note that this will work if the idle interval is not "immediate" (eg. > 1min) because the middleware doesn't guarantee that the connection has been closed and the response was delivered.

**Reply by benallfree** [reply]
Following up again. This works really well:

```go
		app.OnBeforeServe().Add(func(e *core.ServeEvent) error {

			e.Server.IdleTimeout = timeout

			util.Log("Server idle timeout is %d", e.Server.IdleTimeout)
			util.Log("Watchdog timeout is %dms", timeout.Milliseconds())

			var (
				mu               sync.Mutex
				connectionsCount int
				shutdownTimer    *time.Timer
			)

			e.Server.ConnState = func(c net.Conn, cs http.ConnState) {
				mu.Lock()
				util.Log("State is %s", cs)
				defer mu.Unlock()

				switch cs {
				case http.StateNew:
					// Cancel any ongoing shutdown timer
					if shutdownTimer != nil {
						shutdownTimer.Stop()
					}
					connectionsCount++
					util.Log("Connections: %d", connectionsCount)
				case http.StateClosed, http.StateHijacked:
					connectionsCount--
					util.Log("Connections: %d", connectionsCount)

					// If no active connections, start the shutdown timer
					if connectionsCount == 0 {
						if shutdownTimer != nil {
							shutdownTimer.Stop()
						}
						shutdownTimer = time.AfterFunc(timeout, func() {
							util.Log("Closing server connection: %d", connectionsCount)

							e.Server.Close() // gracefully shut down the server
						})
					}
				}
			}

			return nil
		})
```

**Reply by benallfree** [reply]
@ganigeorgiev do you see any reason this same logic couldn't be implemented as a pb_hook?

**Reply by ganigeorgiev** [reply]
@benallfree The goja runtime is not goroutine safe and there are no concurrency primitives exposed to the `pb_hooks` (_also due to the way how our integration works we don't have an event loop so setInterval/setTimeout doesn't work_).

**Comment by benallfree** [comment]
@ganigeorgiev Thinking more about this... is there any reason I need to spawn a separate `pocketbase` process if they have no `pb_hooks`? Nobody is allowed to have a custom binary of `pocketbase`, so we are just concerned about JS hooks. I think PocketHost would be dramatically more efficient if it could spawn a PocketBase server as a goroutine.

**Reply by ganigeorgiev** [reply]
Users can execute custom code also via `pb_migrations`.
While it may be more resource efficient to have a single process firing all PocketBase instances, it will be also a lot easier this way to take down (or at least slow) the entire server in case one of the users decide to abuse/stresstest your platform.
Having a dedicated process per instance/user in a vm/container in my opinion is better because you'll be able to apply individual limits over the cpu, memory, storage and bandwidth.

**Comment by benallfree** [comment]
@ganigeorgiev Is there any API endpoint to call to see if a backup is active?

**Reply by ganigeorgiev** [reply]
May not be exactly what you are looking for, but `GET /api/health` (or `pb.health.check()`) returns a `data.canBackup` prop that indicates whether a backup can be performed (_I'll update the API reference docs sometime later today with an example response_).

If it is `false`, then a backup, restore or any other operation that could prevent safely creating a backup is active.
